L02 Exploring Deep Learning Tools: A No-Code Introduction to TensorFlow and Keras

**Introduction**

We read the introduction and understood that the main goal of this lab is to learn how deep learning and pre-trained models work using a no-code approach with Google Colab and VGG16.

In [1]:
import tensorflow as tf
from tensorflow.keras.applications.vgg16 import VGG16, preprocess_input, decode_predictions
from tensorflow.keras.preprocessing import image
import numpy as np

model = VGG16(weights='imagenet')
model.summary()

553467096/553467096 ━━━━━━━━━━━━━━━━━━━━ 6s 0us/step


Model: "vgg16"

┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━┓
┃ Layer (type)                    ┃ Output Shape           ┃       Param # ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━┩
│ input_layer (InputLayer)        │ (None, 224, 224, 3)    │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ block1_conv1 (Conv2D)           │ (None, 224, 224, 64)   │         1,792 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ block1_conv2 (Conv2D)           │ (None, 224, 224, 64)   │        36,928 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ block1_pool (MaxPooling2D)      │ (None, 112, 112, 64)   │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ block2_conv1 (Conv2D)           │ (None, 112, 112, 128)  │        73,856 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ block2_conv2 (Conv2D)           │ (None, 112, 112, 128)  │       147,584 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ block2_pool (MaxPooling2D)      │ (None, 56, 56, 128)    │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ block3_conv1 (Conv2D)           │ (None, 56, 56, 256)    │       295,168 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ block3_conv2 (Conv2D)           │ (None, 56, 56, 256)    │       590,080 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ block3_conv3 (Conv2D)           │ (None, 56, 56, 256)    │       590,080 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ block3_pool (MaxPooling2D)      │ (None, 28, 28, 256)    │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ block4_conv1 (Conv2D)           │ (None, 28, 28, 512)    │     1,180,160 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ block4_conv2 (Conv2D)           │ (None, 28, 28, 512)    │     2,359,808 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ block4_conv3 (Conv2D)           │ (None, 28, 28, 512)    │     2,359,808 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ block4_pool (MaxPooling2D)      │ (None, 14, 14, 512)    │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ block5_conv1 (Conv2D)           │ (None, 14, 14, 512)    │     2,359,808 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ block5_conv2 (Conv2D)           │ (None, 14, 14, 512)    │     2,359,808 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ block5_conv3 (Conv2D)           │ (None, 14, 14, 512)    │     2,359,808 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ block5_pool (MaxPooling2D)      │ (None, 7, 7, 512)      │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ flatten (Flatten)               │ (None, 25088)          │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ fc1 (Dense)                     │ (None, 4096)           │   102,764,544 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ fc2 (Dense)                     │ (None, 4096)           │    16,781,312 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ predictions (Dense)             │ (None, 1000)           │     4,097,000 │
└─────────────────────────────────┴────────────────────────┴───────────────┘

 Total params: 138,357,544 (527.79 MB)

 Trainable params: 138,357,544 (527.79 MB)

 Non-trainable params: 0 (0.00 B)

**Image Preprocessing**

This function gets the image ready to be used by VGG16. It changes the image size to 224 × 224, converts it into an array, and applies the preprocessing needed by the model before making a prediction.


In [2]:
from tensorflow.keras.utils import load_img, img_to_array

def prepare_image(img_path):
    img = load_img(img_path, target_size=(224, 224))
    x = img_to_array(img)
    x = np.expand_dims(x, axis=0)
    x = preprocess_input(x)
    return x

**Making a Prediction**

Here, I use the prepared image as input for VGG16. The model analyzes the image and gives the three most likely predictions with their confidence scores.


In [5]:

img_path = '/content/Steer.jpg'
image_data = prepare_image(img_path)
preds = model.predict(image_data)

results = decode_predictions(preds, top=3)[0]
for i, (imagenet_id, label, score) in enumerate(results):
    print(f"{i+1}. {label}: {score*100:.2f}%")

1/1 ━━━━━━━━━━━━━━━━━━━━ 1s 917ms/step
1. ox: 46.60%
2. bison: 19.04%
3. hartebeest: 11.55%


**Uploading an Image**

This cell lets me upload an image from my computer. After the upload, the image is prepared and sent to VGG16, which then shows the three predictions with the highest confidence.


In [6]:
from google.colab import files
from tensorflow.keras.utils import load_img, img_to_array
import numpy as np

uploaded = files.upload()

for fn in uploaded.keys():
    img_path = fn
    image_data = prepare_image(img_path)
    preds = model.predict(image_data)

    results = decode_predictions(preds, top=3)[0]
    print(f"\n--- Results for {fn} ---")
    for i, (imagenet_id, label, score) in enumerate(results):
        print(f"{i+1}. {label}: {score*100:.2f}%")

Saving dog.jpg to dog.jpg
1/1 ━━━━━━━━━━━━━━━━━━━━ 1s 560ms/step

--- Results for dog.jpg ---
1. golden_retriever: 88.26%
2. otterhound: 1.65%
3. kuvasz: 1.44%


**Modifying the Image**

In this step, I slightly shift the image before sending it to VGG16 again. I then compare the new predictions with the original results to see if a small change in the image affects the model's confidence.


In [10]:
import matplotlib.pyplot as plt

from tensorflow.keras.preprocessing.image import random_shift, random_rotation

img_to_test = load_img(fn, target_size=(224, 224))
x_orig = img_to_array(img_to_test)

x_modified = random_shift(x_orig,0.2,0.2,row_axis=0,col_axis=1,channel_axis=2)
x_input = np.expand_dims(x_modified, axis=0)
x_input = preprocess_input(x_input)

preds_mod = model.predict(x_input)
results_mod = decode_predictions(preds_mod, top=3)[0]

print(f"\n--- Results after modifying image (Step 5) ---")
for i, (imagenet_id, label, score) in enumerate(results_mod):
    print(f"{i+1}. {label}: {score*100:.2f}%")

1/1 ━━━━━━━━━━━━━━━━━━━━ 1s 564ms/step

--- Results after modifying image (Step 5) ---
1. golden_retriever: 86.04%
2. otterhound: 2.78%
3. Labrador_retriever: 1.72%


**Pixel Shift with NumPy**

Here, I use NumPy to move the image pixels 15 positions to the right. The modified image is then sent to VGG16 again so I can compare the prediction with the original image.


In [11]:
import numpy as np

x_modified = np.roll(x_orig, shift=15, axis=1)
x_input = np.expand_dims(x_modified, axis=0)
x_input = preprocess_input(x_input)

preds_mod = model.predict(x_input)
results_mod = decode_predictions(preds_mod, top=3)[0]

print(f"\n--- Results after modifying image (Step 5) ---")
for i, (imagenet_id, label, score) in enumerate(results_mod):
    print(f"{i+1}. {label}: {score*100:.2f}%")

1/1 ━━━━━━━━━━━━━━━━━━━━ 1s 562ms/step

--- Results after modifying image (Step 5) ---
1. golden_retriever: 86.44%
2. otterhound: 1.82%
3. kuvasz: 1.23%


**Conclusion and Discussion**

In this lab, I learned how to use the pre-trained VGG16 model to classify images. I learned that the image needs to be resized and preprocessed before making a prediction.

My dog image was classified as a Golden Retriever with 88.26% confidence. After modifying the image, the confidence changed to 86.04% and then 86.44%, while Golden Retriever remained the top prediction. This showed me that small changes to an image can affect the model's confidence.

I also learned how to fix errors during the lab, including the image download error and the random_shift function error. Overall, this lab helped me better understand image classification and how pre-trained deep learning models work.

**Feedback and Assessment**

I found this lab helpful because I could work with a real pre-trained deep learning model without building one from the beginning. I learned how to prepare an image, make predictions with VGG16, and see how small changes to an image can affect the results.

I also learned from the errors I had during the lab and how to make the code work with my version of Keras. Overall, I have a better understanding of how VGG16 can be used for image classification.
